src/core/reference/include/openvino/reference/floor_mod.hpp

```diff
@@ -8,22 +8,39 @@
 #include <cstddef>
 
 #include "openvino/reference/autobroadcast_binop.hpp"
+#include "openvino/reference/mod.hpp"
 
 namespace ov {
 namespace reference {
+namespace func {
+
+template <class T, typename std::enable_if<std::is_unsigned<T>::value>::type* = nullptr>
+constexpr T floor_mod(const T x, const T y) {
+    return mod(x, y);
+}
+
+template <class T, typename std::enable_if<std::is_integral<T>::value && std::is_signed<T>::value>::type* = nullptr>
+T floor_mod(const T x, const T y) {
+    // Cast to double is needed for integer input,
+    // otherwise std::floor will act like std::trunc
+    const auto floor_div = static_cast<T>(std::floor(x / static_cast<double>(y)));
+    return x - (floor_div * y);
```

Comment:
looks like there may be an overflow during multiplication versus previous implementation because `floor_div` is casted to T instead of `double`. 

For example, in case of T == float, `y * std::floor(x / divisor)` will be computed in doubles, but in your implementation `floor_div * y` will be in floats

Reply:
This function is only used for integral signed and there is one situation where overflow can be `(x = min<T>(), y = -1)`.
The result e.g. for `int8_t` inputs `(-128, -1)` then floored division will result with 128 which after cast is `-128`. But this is ok as will result as modulo `0` which is correct for `y=-1` (same will be for other ints). in other cases the division is always in int range so multiplication will be also in range.

But changed to previous implementation to have common for signed values (integral and floating-point).